core/range/minmax_spec.rb

```diff
@@ -0,0 +1,66 @@
+require_relative '../../spec_helper'
+
+describe 'Range#minmax' do
+  before(:each) do
+    @x = mock('x')
+    @y = mock('y')
+
+    @x.should_receive(:<=>).with(@y).any_number_of_times.and_return(-1)
+    @x.should_receive(:<=>).with(@x).any_number_of_times.and_return(0)
+    @y.should_receive(:<=>).with(@x).any_number_of_times.and_return(1)
+    @y.should_receive(:<=>).with(@y).any_number_of_times.and_return(0)
+  end
+
+  ruby_version_is '2.7' do # TODO: Remove this before merging!
+    class Range
+      # This monkey patch roughly matches the implementation of the bug fix in MRI's range.c
+      def minmax
+        return super if block_given?
+
+        [min, max]
+      end
+    end
+  end
+
+  describe 'by enumerating the range' do
+    before(:each) do
+      @x.should_receive(:succ).once.and_return(@y)
+    end
+
+    ruby_version_is ''...'2.7' do
+      it 'should return the minimum and maximum elements' do
+        (@x..@y).minmax.should == [@x, @y]
+      end
```

Comment:
What happens for this spec on 2.7?
Also what happens if it's an exclusive Range?

Reply:
The spec exists in two sections. This section is the one where calling `@x.succ` is allowed, which is what happens pre-2.7. The other section disallows calling `@x.succ`, describing the 2.7 behaviour of computing `minmax` without iterating over the range.

Effectively, the way I organized it was

```ruby
describe 'by iterating the range' do
  before(:each, &:allow_range_iteration)

  it_returns_min_and_max if ruby_version_is ''...'2.7'

  it_should_do_other_things_that_need_to_iterate
end

describe 'without iterating the range' do
  before(:each, &:prevent_range_iteration)

  it_returns_min_and_max if ruby_version_is '2.7'

  it_should_do_other_things_that_skip_iteration
end
```

It's the only thing that changed in 2.7, so I was trying to figure out a good way to capture the change while also keeping things concise.

Reply:
I find the structure a bit confusing here. Typically the < 2.7 and 2.7 cases should be next to each other so we can easily compare before/after 2.7.

How about having `@x_iterating, @y_iterating, @x_not_iterating, @y_not_iterating` or so, and then we can have everything in a single describe? Then each `it` can mention `... by/without enumerating the range`, which is some duplication but I don't think that matters much compared to the improved readability.

The interesting part is < 2.7
`should return the minimum and maximum elements by enumerating the range`
and >= 2.7
`should return the minimum and maximum elements without enumerating the range`

Reply:
Yeah, upon going through them again I realized I broken them apart in the wrong way. I saw the repetition in the `it` and figured I'd extract that.

I think the way I'll refactor it is to extract `describe` blocks for the _kinds_ of ranges (inclusive vs exclusive), which should then make it a lot clearer that the new intended behaviour has changed only for inclusive ranges, and that the bug exists only for exclusive ranges.